go-controller/pkg/node/gateway_localnet.go

```diff
@@ -151,92 +232,185 @@ func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFa
-		IPAddresses:    []*net.IPNet{gatewayIPCIDR},
-		NextHops:       []net.IP{gatewayNextHop},
+		IPAddresses:    ipAddrs,
+		NextHops:       nextHops,
 		NodePortEnable: config.Gateway.NodeportEnable,
 	})
+
+}
+
+func addLinkAddresses(link netlink.Link, ipData []*localnetData) error {
+
+	// First flush addressed
+	err := flushLinkAddress(link)
 	if err != nil {
 		return err
 	}
 
-	if utilnet.IsIPv6CIDR(subnet) {
-		// TODO - IPv6 hack ... for some reason neighbor discovery isn't working here, so hard code a
-		// MAC binding for the gateway IP address for now - need to debug this further
-		err = util.LinkNeighAdd(link, gatewayIP, macAddress)
-		if err == nil {
-			klog.Infof("Added MAC binding for %s on %s", gatewayIP, localnetGatewayNextHopPort)
-		} else {
-			klog.Errorf("Error in adding MAC binding for %s on %s: %v", gatewayIP, localnetGatewayNextHopPort, err)
+	// Now add all addresses
+	for _, data := range ipData {
+		err := util.LinkAddrAdd(link, data.gatewayNextHopCIDR)
+		if err != nil {
+			return err
 		}
 	}
 
-	ipt, err := localnetIPTablesHelper(subnet)
-	if err != nil {
-		return err
+	return nil
+}
+
+func flushLinkAddress(link netlink.Link) error {
+
+	return util.LinkAddrFlush(link)
+
+}
+
+type localnetData struct {
+	ipVersion                         int
+	ipt                               util.IPTablesHelper
+	gatewayIP, gatewayNextHop         net.IP
+	gatewaySubnetMask                 net.IPMask
+	gatewayIPCIDR, gatewayNextHopCIDR *net.IPNet
+}
+
+// Function that populates releavant IP structures
+
+func evaluateInputNetworks(subnets []*net.IPNet) error {
+
+	if len(subnets) > totalSupportedSubnets {
+		return fmt.Errorf("Number of subnets provided is : %d  which is more than %d", len(subnets), totalSupportedSubnets)
+	}
+	var ipV4NetworkCount, ipV6NetworkCount int
+	for _, subnet := range subnets {
+		if utilnet.IsIPv6CIDR(subnet) {
+			ipV6NetworkCount++
+		} else {
+			ipV4NetworkCount++
+		}
+	}
+	if ipV4NetworkCount > supportedSubnetsPerFamily || ipV6NetworkCount > supportedSubnetsPerFamily {
+		return fmt.Errorf("Only one IPv4 and/or IPv6 subnets can be specified. Specified %d IP v4 and %d "+
+			"IP v6 subnets", ipV4NetworkCount, ipV6NetworkCount)
 	}
+	return nil
+}
 
-	err = localnetGatewayNAT(ipt, localnetGatewayNextHopPort, gatewayIP)
-	if err != nil {
-		return fmt.Errorf("Failed to add NAT rules for localnet gateway (%v)", err)
+func constructIPData(subnets []*net.IPNet) ([]*localnetData, error) {
+	netData := make([]*localnetData, 0)
+	for _, subnet := range subnets {
+		if utilnet.IsIPv6CIDR(subnet) {
+			data, err := newIPV6Data()
+			if err != nil {
+				return nil, err
+			}
+			netData = append(netData, data)
+		} else {
+			data, err := newIPV4Data()
+			if err != nil {
+				return nil, err
+			}
+			netData = append(netData, data)
+		}
 	}
+	return netData, nil
+}
 
-	if config.Gateway.NodeportEnable {
-		err = localnetNodePortWatcher(ipt, wf, gatewayIP)
+func newIPV4Data() (*localnetData, error) {
```

Comment:
`newIPV4Data` and `newIPV6Data` can just be merged into `constructIPData`

Reply:
Since for each IP family there are separate IP address, next hop and IPT objects , wanted to separate them this way so that it's clear and easier on the eyes and also easier to write a unit test on the method when we get there. I would prefer to keep it this way if that's okay with you. I'll address the inconsistencies in the naming conventions as per your suggestion.